pkg/portal/util/clientcache/clientcache.go

```diff
@@ -0,0 +1,76 @@
+package clientcache
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"net/http"
+	"sync"
+	"time"
+)
+
+// ClientCache is a cache for *http.Clients.  It allows us to reuse clients and
+// connections across multiple incoming calls, saving us TCP, TLS and proxy
+// initialisations.
+type ClientCache interface {
+	Get(interface{}) *http.Client
+	Put(interface{}, *http.Client)
+}
+
+type clientCache struct {
+	mu  sync.Mutex
+	now func() time.Time
+	ttl time.Duration
+	m   map[interface{}]*v
+}
+
+type v struct {
+	expires time.Time
+	cli     *http.Client
+}
+
+// New returns a new ClientCache
+func New(ttl time.Duration) ClientCache {
+	return &clientCache{
+		now: time.Now,
+		ttl: ttl,
+		m:   map[interface{}]*v{},
+	}
+}
+
+// call holding c.mu
+func (c *clientCache) expire() {
+	now := c.now()
+	for k, v := range c.m {
+		if now.After(v.expires) {
+			v.cli.CloseIdleConnections()
+			delete(c.m, k)
+		}
+	}
+}
+
```

Comment:
Does this implies that if we dont call any Get, Put, clients will not be expired? Potentially sitting inmemory?

Reply:
Yes.  Do you think that's adequate?

Reply:
It does the job. I just never seen an implementation like this. Usually, its some goroutine running and checking expiring things. Basically, I'm worried if somebody can access the memory of the VM (one more side-channel or smth) we could have stale clients in regions where we run 1-2 clusters for months... But at the same time if somebody dumped the memory - we have bigger issues :man_shrugging:  I think I'm fine with this as long as we are aware